iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+{
+  (void)event;
+  (void)e;
+
+  // Schedule up the next partition
+  if (partition < cc->partition_count())
+    mutex = cc->get_partition(partition).lock.get();
+  else
+    mutex = cont->mutex;
+  SET_HANDLER((CacheSyncHandler)&RefCountCacheSync::copy_partition);
+  e->schedule_imm(ET_TASK);
+  return EVENT_CONT;
+}
+
+// Open the tmp file, etc.
+template <class C>
+int
+RefCountCacheSync<C>::initialize_storage(int event, Event *e)
+{
+  (void)event;                                                                                        // unused
+  this->fd = this->socket_manager.open(this->tmp_filename.c_str(), O_TRUNC | O_RDWR | O_CREAT, 0644); // TODO: configurable perms
+
+  if (this->fd <= 0) {
+    Warning("Unable to create temporary file %s, unable to persist hostdb: %d\n", this->tmp_filename.c_str(), this->fd);
+    cont->handleEvent(REFCOUNT_CACHE_EVENT_SYNC, 0);
+    delete this;
+    return EVENT_DONE;
+  }
+
+  // Write out the header
+  int ret = this->write_to_disk((char *)&this->cc->get_header(), sizeof(RefCountCacheHeader));
+  if (ret < 0) {
+    Warning("Error writing cache header to %s: %d", this->tmp_filename.c_str(), ret);
+    cont->handleEvent(REFCOUNT_CACHE_EVENT_SYNC, 0);
+    delete this;
+    return EVENT_DONE;
+  }
+
+  SET_HANDLER((CacheSyncHandler)&RefCountCacheSync::copy_partition);
+  e->schedule_imm(ET_TASK);
+  return EVENT_CONT;
+}
+
+// do the final mv and close of file handle
+template <class C>
+int
+RefCountCacheSync<C>::finalize_sync()
+{
+  // fsync the fd we have
+  int fsync_ret = this->socket_manager.fsync(this->fd);
+  if (fsync_ret != 0) {
+    return fsync_ret;
+  }
+
+  int dir_fd = this->socket_manager.open(this->dirname.c_str(), O_DIRECTORY); // Correct permissions?
+  if (dir_fd <= 0) {
+    return dir_fd;
+  }
+  // move the file
+  int ret = rename(this->tmp_filename.c_str(), this->filename.c_str());
+
+  if (ret != 0) {
+    return ret;
+  }
+
+  // fsync the dir
+  int fsync_dir_ret = this->socket_manager.fsync(dir_fd);
+  if (fsync_dir_ret != 0) {
+    return fsync_dir_ret;
+  }
+
+  int dir_close_ret = this->socket_manager.close(dir_fd);
+  if (dir_close_ret != 0) {
+    return dir_close_ret;
+  }
+
+  int close_ret = this->socket_manager.close(this->fd);
+  if (close_ret != 0) {
+    return close_ret;
+  }
+
+  if (this->rsb) {
+    RecSetRawStatCount(this->rsb, refcountcache_last_sync_time, Thread::get_hrtime() / HRTIME_SECOND);
+    RecSetRawStatCount(this->rsb, refcountcache_last_total_items, this->total_items);
+    RecSetRawStatCount(this->rsb, refcountcache_last_total_size, this->total_size);
+  }
+
+  return 0;
+}
+
+// Write *i to this->fd, if there is an error we'll just stop this continuation
+// TODO: reschedule the continuation if the disk was busy?
+template <class C>
+int
+RefCountCacheSync<C>::write_to_disk(char *i, int size)
+{
+  int written = 0;
+  while (written < size) {
+    int ret = this->socket_manager.write(this->fd, i + written, size - written);
```

Comment:
From the man page to write:
`On  success,  the  number  of bytes written is returned (zero indicates
       nothing was written).  On error, -1  is  returned,  and  errno  is  set
       appropriately.
`

Meaning that there are no gaurantees that all the bytes you wanted to write are written out to disk-- so we have to check how many where written, and if it wasn't all of them then we need to call write again with the remaining bytes. IMO this would probably be a useful thing in `SocketManager` as well.